File: app/lib/models/community.dart
/// An open, followable community (specs/api/communities.md).
class Community {
  const Community({
    required this.id,
    required this.name,
    this.tagline,
    this.icon,
    this.color,
    this.photo,
    this.followerCount = 0,
    this.youFollow = false,
    this.yourRole,
  });

  final String id;
  final String name;
  final String? tagline;
  final String? icon;
  final String? color;
  final String? photo;
  final int followerCount;
  final bool youFollow;

  /// `'leader'` when the caller may post/edit this community's events, else null.
  final String? yourRole;

  bool get isLeader => yourRole == 'leader';

  factory Community.fromJson(Map<String, dynamic> json) => Community(
    id: json['id'] as String,
    name: json['name'] as String? ?? '',
    tagline: json['tagline'] as String?,
    icon: json['icon'] as String?,
    color: json['color'] as String?,
    photo: json['photo'] as String?,
    followerCount: (json['follower_count'] as num?)?.toInt() ?? 0,
    youFollow: json['you_follow'] as bool? ?? false,
    yourRole: json['your_role'] as String?,
  );
}

/// A leader-broadcast, born-confirmed recurring event + the viewer's RSVP.
class CommunityEvent {
  const CommunityEvent({
    required this.id,
    required this.title,
    this.communityName,
    this.communityIcon,
    this.activityTypeLabel,
    this.time,
    this.recurrence,
    this.location,
    this.goingCount = 0,
    this.publicGoing = const [],
    this.youGoing = false,
    this.youPublic = false,
    this.threadCount = 0,
  });

  final String id;
  final String title;
  final String? communityName;
  final String? communityIcon;
  final String? activityTypeLabel;
  final String? time;
  final String? recurrence;
  final String? location;
  final int goingCount;
  final List<String> publicGoing; // first names for the face-pile
  final bool youGoing;
  final bool youPublic;
  final int threadCount;

  factory CommunityEvent.fromJson(Map<String, dynamic> json) {
    final community = json['community'] as Map<String, dynamic>?;
    final type = json['activity_type'] as Map<String, dynamic>?;
    final rsvp = json['your_rsvp'] as Map<String, dynamic>? ?? const {};
    return CommunityEvent(
      id: json['id'] as String,
      title: json['title'] as String? ?? '',
      communityName: community?['name'] as String?,
      communityIcon: community?['icon'] as String?,
      activityTypeLabel: type?['label'] as String?,
      time: json['time'] as String?,
      recurrence: json['recurrence'] as String?,
      location: json['location'] as String?,
      goingCount: (json['going_count'] as num?)?.toInt() ?? 0,
      publicGoing: ((json['public_going'] as List<dynamic>?) ?? const [])
          .map(
            (e) => (e as Map<String, dynamic>)['first_name'] as String? ?? '',
          )
          .where((s) => s.isNotEmpty)
          .toList(),
      youGoing: rsvp['going'] as bool? ?? false,
      youPublic: rsvp['public'] as bool? ?? false,
      threadCount: (json['thread_count'] as num?)?.toInt() ?? 0,
    );
  }
}

File: server/src/contracts/community.ts
import type { CommunitySummary, EventView } from '../domain/community/service.ts'

// Community wire shapes (specs/api/communities.md).
export function serializeCommunity(s: CommunitySummary) {
  return {
    id: s.community.id,
    name: s.community.name,
    tagline: s.community.tagline,
    icon: s.community.icon,
    color: s.community.color,
    photo: s.community.photo,
    follower_count: s.followerCount,
    you_follow: s.youFollow,
    your_role: s.yourRole === 'leader' ? 'leader' : null,
  }
}

export function serializeCommunityEvent(
  v: EventView,
  community: { id: string; name: string; icon: string | null; color: string | null },
) {
  const c = v.event
  return {
    id: c.id,
    community: {
      id: community.id,
      name: community.name,
      icon: community.icon,
      color: community.color,
    },
    title: c.title,
    activity_type: v.activityTypeLabel
      ? { id: c.activityTypeId, label: v.activityTypeLabel }
      : null,
    time: c.time,
    recurrence: c.recurrence,
    location: c.location,
    going_count: v.goingCount,
    public_going: v.publicGoing.map((p) => ({
      id: p.id,
      first_name: p.firstName,
      photo: p.photo,
    })),
    your_rsvp: v.yourRsvp,
    thread_count: v.threadCount,
  }
}
